### PR TITLE
[neox-2.x] add interop service

### PR DESCRIPTION
**Pull request by KickSeason**
### Step 3/6  from `neo-ngd/neo` branch `neox-2.x`
Add 1 new interop services
* `Neo.Cryptography.Keccak256`

@erikzhang @shargon  Please take a look.

### CODE REVIEW

File: neo/Cryptography/ECC/ECDsa.cs [hunk 1]
```diff
@@ -104,5 +104,89 @@ public bool VerifySignature(byte[] message, BigInteger r, BigInteger s)
...
+                W = ECPoint.Multiply(t1, invr);
+            }
+            var test = W;
+            return Q;
+        }
+
+        public static ECPoint KeyRecover(ECCurve curve, BigInteger r, BigInteger s, byte[] msg, bool IsEven, bool doChecks)
+        {
+            // calculate h
+            BigInteger h = (curve.Q + 1 + 2 * (BigInteger)Math.Sqrt((double)curve.Q)) / curve.N;
```
Comment (shargon) [inline neo/Cryptography/ECC/ECDsa.cs:151]: it's 1 in both curves?

File: neo/Cryptography/Helper.cs [hunk 2]
```diff
@@ -91,6 +92,15 @@ public static string Base58CheckEncode(this byte[] data)
             return Base58.Encode(buffer);
         }
 
+        public static byte[] Keccak256(this byte[] value)
+        {
+            var digest = new KeccakDigest(256);
```
[the remark this answers is not in the document]
Comment (KickSeason) [reply neo/Cryptography/Helper.cs:97]: It's for ethereum.  when verify cross chain tx and block header from ethereum, we need this algorithm.

[the remark this answers is not in the document]
Comment (Lichen9618) [reply neo/Cryptography/Helper.cs:97]: Keccak256 algorithm is necessary in crosschain project, due to it is used in ETH signatures verification.

[the remark this answers is not in the document]
Comment (wanglongfei88) [reply neo/Cryptography/Helper.cs:97]: We add this algorithm so it is more compatible with cross chain tx and block header, signature verification with other blockchain, like Ethereum. 

It is essential for crosschain solution one Neo2 as well as Neo3.

File: neo/SmartContract/NeoService.cs [hunk 3]
```diff
@@ -165,6 +169,36 @@ public NeoService(TriggerType trigger, Snapshot snapshot)
...
+        private bool Ecrecover(ExecutionEngine engine)
+        {
+            var r = new System.Numerics.BigInteger(engine.CurrentContext.EvaluationStack.Pop().GetByteArray().Reverse().ToArray());
+            var s = new System.Numerics.BigInteger(engine.CurrentContext.EvaluationStack.Pop().GetByteArray().Reverse().ToArray());
+            bool v = engine.CurrentContext.EvaluationStack.Pop().GetBoolean();
+            byte[] messageHash = engine.CurrentContext.EvaluationStack.Pop().GetByteArray();
+            try
+            {
+                ECPoint point = ECDsa.KeyRecover(ECCurve.Secp256k1, r, s, messageHash, v, true);
+                engine.CurrentContext.EvaluationStack.Push(point.EncodePoint(false).Skip(1).ToArray());//加120504
```
[the remark this answers is not in the document]
Comment (Lichen9618) [reply neo/SmartContract/NeoService.cs:186]: This is a compressed flag in algorithm. But it is not exist in ETH's pubkey. so we remove it for compatiable

**Comment by KickSeason** [comment]
@shargon I removed `ECRecover`. Will commit another pr later.

**Comment by KickSeason** [comment]
@erikzhang Can you take a look?

File: neo/Cryptography/Helper.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shargon) [reply neo/Cryptography/Helper.cs:97]: I agree, but for neo 3, for neo2 I think that we should focus the cross chain only in be compatible with neo3.

[the remark this answers is not in the document]
Comment (shargon) [reply neo/Cryptography/Helper.cs:97]: But if you think that it will be easier, in order to duplicate the logic later, i am not against to.

**Comment by KickSeason** [comment]
I close this. I can reopen when we decide whether NEO should support other popular signature algorithm.

**Pull request closed without merging**